src/main/java/com/minecolonies/structures/helpers/Structure.java

```diff
@@ -84,61 +98,312 @@
-                }
-                else
-                {
-                    decorationFolder = new File(FMLCommonHandler.instance().getMinecraftServerInstance().getDataDirectory(), "minecolonies/");
-                }
-                if(decorationFolder.exists())
+        if (inputStream == null)
+        {
+            inputStream = Structure.getStream(correctStructureName);
+        }
+
+        if (inputStream == null)
+        {
+            Log.getLogger().warn(String.format("Failed to load template %s", correctStructureName));
+            return;
+        }
+
+        try
+        {
+            this.md5 = Structure.calculateMD5(Structure.getStream(correctStructureName));
+            this.template = readTemplateFromStream(inputStream);
+        }
+        catch (final IOException e)
+        {
+            Log.getLogger().warn(String.format("Failed to load template %s", correctStructureName), e);
+        }
+        finally
+        {
+            IOUtils.closeQuietly(inputStream);
+        }
+    }
+
+    /**
+     * Get the file representation of the cached schematics' folder.
+     *
+     * @return the folder for the cached schematics
+     */
+    @Nullable
+    public static File getCachedSchematicsFolder()
+    {
+        if (FMLCommonHandler.instance().getMinecraftServerInstance() == null)
+        {
+            if (ColonyManager.getServerUUID() != null)
+            {
+                return new File(Minecraft.getMinecraft().mcDataDir, Constants.MOD_ID + "/" + ColonyManager.getServerUUID());
+            }
+            else
+            {
+                Log.getLogger().error("ColonyManager.getServerUUID() => null this should not happen");
+                return null;
+            }
+        }
+        return new File(FMLCommonHandler.instance().getMinecraftServerInstance().getEntityWorld().getSaveHandler().getWorldDirectory()
+                          + "/" + Constants.MOD_ID);
+    }
+
+    /**
+     * get the schematic folder for the client.
+     *
+     * @return the client folder.
+     */
+    public static File getClientSchematicsFolder()
+    {
+        return new File(Minecraft.getMinecraft().mcDataDir, Constants.MOD_ID);
+    }
+
+    /**
+     * get a InputStream for a give structureName.
+     * <p>
+     * Look into the following director (in order):
+     * - scan
+     * - cache
+     * - schematics folder
+     * - jar
+     * It should be the exact oppsite that the way used to buikd the list.
+     *
+     * @param structureName name of the structure to load
+     * @return the input stream or null
+     */
+    @Nullable
+    public static InputStream getStream(final String structureName)
+    {
+        final Structures.StructureName sn = new Structures.StructureName(structureName);
+        InputStream inputstream = null;
+        if (Structures.SCHEMATICS_CACHE.equals(sn.getPrefix()))
+        {
+            return Structure.getStreamFromFolder(Structure.getCachedSchematicsFolder(), structureName);
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER') Close this "InputStream". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)